storage/aptosdb/src/pruner/mod.rs

```diff
@@ -37,50 +33,95 @@ use std::{
+}
+
+#[cfg(test)]
+pub enum PrunerIndex {
+    StateStorePrunerIndex,
+    TransactionStorePrunerIndex,
 }
 
 impl Pruner {
     /// Creates a worker thread that waits on a channel for pruning commands.
-    pub fn new(db: Arc<DB>, historical_versions_to_keep: u64) -> Self {
+    pub fn new(
+        db: Arc<DB>,
+        storage_pruner_config: StoragePrunerConfig,
+        transaction_store: Arc<TransactionStore>,
+    ) -> Self {
         let (command_sender, command_receiver) = channel();
 
-        let worker_progress = Arc::new(AtomicU64::new(0));
-        let worker_progress_clone = Arc::clone(&worker_progress);
+        let least_readable_version = Arc::new(Mutex::new(vec![0, 0]));
+        let worker_progress_clone = Arc::clone(&least_readable_version);
 
-        DIEM_STORAGE_PRUNE_WINDOW.set(historical_versions_to_keep as i64);
+        DIEM_STORAGE_PRUNE_WINDOW
+            .set(storage_pruner_config.state_store_prune_window.unwrap() as i64);
+        let worker = Worker::new(
+            db,
+            transaction_store,
+            command_receiver,
+            least_readable_version,
+        );
         let worker_thread = std::thread::Builder::new()
             .name("aptosdb_pruner".into())
-            .spawn(move || Worker::new(db, command_receiver, worker_progress_clone).work())
+            .spawn(move || worker.work())
             .expect("Creating pruner thread should succeed.");
 
         Self {
-            historical_versions_to_keep,
+            state_store_prune_window: storage_pruner_config
+                .state_store_prune_window
+                .expect("State store prune window must be specified"),
+            default_prune_window: storage_pruner_config
+                .default_prune_window
+                .expect("Default prune window must be specified"),
             worker_thread: Some(worker_thread),
             command_sender: Mutex::new(command_sender),
-            worker_progress,
+            least_readable_version: worker_progress_clone,
         }
     }
 
+    pub fn get_state_store_pruner_window(&self) -> Version {
+        self.state_store_prune_window
+    }
+
     /// Sends pruning command to the worker thread when necessary.
     pub fn wake(&self, latest_version: Version) {
-        if latest_version > self.historical_versions_to_keep {
-            let least_readable_version = latest_version - self.historical_versions_to_keep;
+        let least_readable_state_store_version = if latest_version > self.state_store_prune_window {
+            latest_version - self.state_store_prune_window
+        } else {
+            0
+        };
+
+        let least_readable_default_store_version = if latest_version > self.state_store_prune_window
+        {
+            latest_version - self.default_prune_window
+        } else {
+            0
+        };
+
+        if least_readable_state_store_version > 0 || least_readable_default_store_version > 0 {
```

Comment:
Using different windows, some pruner will receive 0 before any actual work needs to be done anyway, so maybe just send every time? (So that if there's a problem with the unnecessary commands, we discover it a bit earlier)

Reply:
Sure, removed the `if` statement - that makes it a bit cleaner as well.